src/commands/promote.ts

```diff
@@ -105,102 +102,112 @@ export default class Promote extends Command {
-    const debArtifacts = [
-      templateShortKey('deb', {bin: config.bin, versionShaRevision: debVersion(buildConfig), arch: 'amd64' as any}),
-      templateShortKey('deb', {bin: config.bin, versionShaRevision: debVersion(buildConfig), arch: 'i386' as any}),
-      'Packages.gz',
-      'Packages.xz',
-      'Packages.bz2',
-      'Release',
-      'InRelease',
-      'Release.gpg',
-    ]
-    if (flags.deb) {
-      this.log(`Promoting debian artifacts to ${flags.channel}`)
-      for (const artifact of debArtifacts) {
-        const debCopySource = cloudBucketCommitKey(`apt/${artifact}`)
-        const debKey = cloudChannelKey(`apt/${artifact}`)
-        // eslint-disable-next-line no-await-in-loop
-        await aws.s3.copyObject(
-          {
-            Bucket: s3Config.bucket,
-            ACL: s3Config.acl || 'public-read',
-            CopySource: debCopySource,
-            Key: debKey,
-            CacheControl: maxAge,
-            MetadataDirective: 'REPLACE',
-          },
+    const maybePromoteDeb = async () => {
+      // copy debian artifacts
+      const debArtifacts = [
+        templateShortKey('deb', {bin: config.bin, versionShaRevision: debVersion(buildConfig), arch: 'amd64' as any}),
+        templateShortKey('deb', {bin: config.bin, versionShaRevision: debVersion(buildConfig), arch: 'i386' as any}),
+        'Packages.gz',
+        'Packages.xz',
+        'Packages.bz2',
+        'Release',
+        'InRelease',
+        'Release.gpg',
+      ]
+      if (flags.deb) {
+        this.log(`Promoting debian artifacts to ${flags.channel}`)
+        await Promise.all(debArtifacts.flatMap(artifact => {
+          const debCopySource = cloudBucketCommitKey(`apt/${artifact}`)
+          const debKey = cloudChannelKey(`apt/${artifact}`)
+          // apt expects ../apt/dists/versionName/[artifacts] but oclif wants varsions/sha/apt/[artifacts]
+          // see https://github.com/oclif/oclif/issues/347 for the AWS-redirect that solves this
+          // this workaround puts the code in both places that the redirect was doing
+          // with this, the docs are correct. The copies are all done in parallel so it shouldn't be too costly.
+          const workaroundKey = cloudChannelKey(`apt/./${artifact}`)
+          return [
+            aws.s3.copyObject(
+              {
+                ...awsDefaults,
+                CopySource: debCopySource,
+                Key: debKey,
+              },
+            ),
+            aws.s3.copyObject(
+              {
+                ...awsDefaults,
+                CopySource: debCopySource,
+                Key: workaroundKey,
+              },
+            ),
+          ]
+        }),
+
         )
       }
     }
+
+    await Promise.all(buildConfig.targets.flatMap(target => [
+      promoteManifest(target),
+      promoteGzAndMaybeIndexes(target),
+      maybePromoteXzAndMaybeIndexes(target),
+      maybePromoteMacos(),
+      maybePromoteWin(),
+      maybePromoteDeb(),
+    ]),
```

Comment:
This `Promise.all` combined with a `.flatMap` and functions that _maybe_ do something is very complex and unintuitive.  Can you add some comments to clarify what exacly is going on here? 

It also seems like the functions are being called as part of the `.flatMap`? Shouldn't the functions just be returned so that `Promise.all` can run them? And why are the promote macos, win, and deb functions included in this if they don't depend on a `target` being passed into them? Seems like we're just running them an unnecessary amount of times.

I suggest changing the mac, win, and deb functions to not be "maybe" functions and including them like so:
```suggestion
    const promotionTargets = buildConfig.targets.flatMap(target => [
      promoteManifest(target),
      promoteGzAndMaybeIndexes(target),
      maybePromoteXzAndMaybeIndexes(target)
    })
    if (flags.macos) promotionTargets.push(promoteMacos)
    if (flags.win) promotionTargets.push(promoteWin)
    if (flags.deb) promotionTargets.push(promoteDeb)

    await Promise.all(promotionTargets)
```

Reply:
good feedback on hitting the installers more than once.  I also took the maybe out of the xz for consistency...see if you like that better.